workflow/controller/controller.go

```diff
@@ -973,6 +1019,22 @@ func (wfc *WorkflowController) addWorkflowInformerHandlers(ctx context.Context)
 				DeleteFunc: func(obj interface{}) {
 					// IndexerInformer uses a delta queue, therefore for deletes we have to use this
 					// key function.
+
+					// Remove finalizers if they exist before deletion
+					pods := wfc.kubeclientset.CoreV1().Pods(wfc.GetManagedNamespace())
+					podList, err := pods.List(ctx, metav1.ListOptions{
+						LabelSelector: fmt.Sprintf("%s=%s", common.LabelKeyWorkflow, obj.(*unstructured.Unstructured).GetName()),
+					})
+					if err != nil {
+						log.WithError(err).Error("Failed to list pods")
+					}
+					for _, p := range podList.Items {
+						patch := createFinalizerPodStatusRemovalPatchIfExists(&p)
+						if err := applyPatchIfExists(ctx, pods, p.Name, patch); err != nil {
```

Comment:
We always do these two methods together - I wonder if there should be a convenience function like `enablePodForDeletion()`?

Reply:
I have addressed this in the commit in the following comment.
https://github.com/argoproj/argo-workflows/pull/12413#discussion_r1446980034